Source/gendung.cpp

```diff
@@ -149,22 +149,25 @@ void MakeSpeedCels()
 		for (x = 0; x < MAXDUNX; x++) {
 			for (i = 0; i < blocks; i++) {
 				pMap = &dpiece_defs_map_2[x][y];
-				if (pMap->mt[i]) {
+				mt = pMap->mt[i];
+				if (mt) {
 					level_frame_count[pMap->mt[i] & 0xFFF]++;
-					level_frame_types[pMap->mt[i] & 0xFFF] = pMap->mt[i] & 0x7000;
+					level_frame_types[pMap->mt[i] & 0xFFF] = mt & 0x7000;
 				}
 			}
 		}
 	}
 
 	pFrameTable = (DWORD *)pDungeonCels;
-	nlevel_frames = pFrameTable[0] & 0xFFFF;
+	nDataSize = pFrameTable[0];
+	nlevel_frames = nDataSize & 0xFFFF;
 
 	for (i = 1; i < nlevel_frames; i++) {
+		z = i;
 #ifdef USE_ASM
 		__asm {
 			mov		ebx, pDungeonCels
-			mov		eax, i
+			mov		eax, z

```

Comment:
That in combination with `z = i;` sounds strange for me (to say the least).
I hope future will reveal more natural looking conditions which forced programmers to use separate (shadow) variable `z` instead of just `i`.

Reply:
That's due to VC inline assembly deficiency, not reassigning loop variable to separate one makes it completely stay out of register making iteration pretty inefficient. Inline assembly is very weird thing and was never implemented for x64 by Microsoft for reasons like this probably.

Reply:
Fair enough! Thanks for another mini-tutorial:)